x/mongo/driver/topology/server_test.go

```diff
@@ -54,6 +54,238 @@ func (cncd *channelNetConnDialer) DialContext(_ context.Context, _, _ string) (n
 	return cnc, nil
 }
 
+func TestServerConnectionTimeout(t *testing.T) {
+	testCases := []struct {
```

Comment:
Yes, I do think it's worth testing. You can test the case where the parent context is cancelled by taking advantage of the fact that anonymous functions can access variables outside of their scope:

```
ctx, cancel := context.WithCancel(context.Background()) // parent context passed to connect()
dialer := func(Dialer) Dialer {
    // Inside the returned DialerFunc, call cancel()
}
```

Testing the case where the child context is cancelled is tougher because `context.WithCancel()` is called inside `connect`, so you don't have access to that `context.CancelFunc`. I see two options here:

1. Spawn a goroutine that calls `connection#closeConnectContext` before dialing. This will block until the child context is fully created and will then cancel it. The Dialer/Handshaker can block waiting for the context to expire as is done in other tests.
2. Alternatively, a simpler implementation would be to just return a `context.Canceled` error from the dialer even though no context has actually expired. This is a bit of a hack but should sufficiently mimic the case where the parent context has not expired, so the pool should be cleared.

Reply:
I've added a new test `TestServerConnectionCancellation` that tests context cancellation before dialing.